File: sdk/python/aragora_sdk/namespaces/review_queue.py
"""
Review Queue Namespace API

Provides methods for the PDB (PR intelligence brief) review queue:
- List open PRs eligible for settlement
- Fetch brief content for a specific PR
- Approve / request-changes / defer settlement actions
- Read session stats (streak, decision time, totals)

All mutating actions preserve the settlement gate: the approve and
request-changes endpoints proxy a GitHub review using the caller's
authenticated identity, not automation credentials.
"""

from __future__ import annotations

from typing import TYPE_CHECKING, Any

if TYPE_CHECKING:
    from ..client import AragoraAsyncClient, AragoraClient


class ReviewQueueAPI:
    """
    Synchronous Review Queue API.

    Example:
        >>> client = AragoraClient(base_url="https://api.aragora.ai")
        >>> prs = client.review_queue.list_prs()
        >>> for pr in prs["prs"]:
        ...     print(pr["number"], pr["title"])
    """

    def __init__(self, client: "AragoraClient") -> None:
        self._client = client

    def list_prs(self, *, include_deferred: bool = False) -> dict[str, Any]:
        """List open PRs currently in the review queue.

        Args:
            include_deferred: If True, include PRs that were deferred. Defaults to False.
        """
        params: dict[str, Any] = {}
        if include_deferred:
            params["include_deferred"] = "1"
        return self._client.request("GET", "/api/v1/review-queue/prs", params=params)

    def get_brief(self, pr_number: int) -> dict[str, Any]:
        """Fetch the PDB brief for a specific PR, or 404 if no brief exists."""
        return self._client.request(
            "GET", f"/api/v1/review-queue/prs/{pr_number}/brief"
        )

    def approve(self, pr_number: int, *, note: str | None = None) -> dict[str, Any]:
        """Submit a GitHub APPROVE review for the PR using the caller's identity."""
        body: dict[str, Any] = {}
        if note is not None:
            body["note"] = note
        return self._client.request(
            "POST", f"/api/v1/review-queue/prs/{pr_number}/approve", json=body
        )

    def request_changes(self, pr_number: int, *, reason: str) -> dict[str, Any]:
        """Submit a GitHub REQUEST_CHANGES review for the PR with a required reason."""
        return self._client.request(
            "POST",
            f"/api/v1/review-queue/prs/{pr_number}/request-changes",
            json={"reason": reason},
        )

    def defer(self, pr_number: int) -> dict[str, Any]:
        """Defer the PR locally (hides it from the queue for ~4 hours)."""
        return self._client.request(
            "POST", f"/api/v1/review-queue/prs/{pr_number}/defer"
        )

    def stats(self) -> dict[str, Any]:
        """Fetch session stats: approvals today, median decision time, streak."""
        return self._client.request("GET", "/api/v1/review-queue/stats")


class AsyncReviewQueueAPI:
    """
    Asynchronous Review Queue API.

    Example:
        >>> client = AragoraAsyncClient(base_url="https://api.aragora.ai")
        >>> prs = await client.review_queue.list_prs()
    """

    def __init__(self, client: "AragoraAsyncClient") -> None:
        self._client = client

    async def list_prs(self, *, include_deferred: bool = False) -> dict[str, Any]:
        """List open PRs currently in the review queue."""
        params: dict[str, Any] = {}
        if include_deferred:
            params["include_deferred"] = "1"
        return await self._client.request(
            "GET", "/api/v1/review-queue/prs", params=params
        )

    async def get_brief(self, pr_number: int) -> dict[str, Any]:
        """Fetch the PDB brief for a specific PR, or 404 if no brief exists."""
        return await self._client.request(
            "GET", f"/api/v1/review-queue/prs/{pr_number}/brief"
        )

    async def approve(
        self, pr_number: int, *, note: str | None = None
    ) -> dict[str, Any]:
        """Submit a GitHub APPROVE review for the PR using the caller's identity."""
        body: dict[str, Any] = {}
        if note is not None:
            body["note"] = note
        return await self._client.request(
            "POST", f"/api/v1/review-queue/prs/{pr_number}/approve", json=body
        )

    async def request_changes(self, pr_number: int, *, reason: str) -> dict[str, Any]:
        """Submit a GitHub REQUEST_CHANGES review for the PR with a required reason."""
        return await self._client.request(
            "POST",
            f"/api/v1/review-queue/prs/{pr_number}/request-changes",
            json={"reason": reason},
        )

    async def defer(self, pr_number: int) -> dict[str, Any]:
        """Defer the PR locally (hides it from the queue for ~4 hours)."""
        return await self._client.request(
            "POST", f"/api/v1/review-queue/prs/{pr_number}/defer"
        )

    async def stats(self) -> dict[str, Any]:
        """Fetch session stats: approvals today, median decision time, streak."""
        return await self._client.request("GET", "/api/v1/review-queue/stats")

File: sdk/typescript/src/namespaces/review-queue.ts
/**
 * Review Queue Namespace API
 *
 * Methods for the PDB (PR intelligence brief) review queue: listing open PRs
 * eligible for settlement, fetching briefs, submitting approve /
 * request-changes reviews (via the caller's authenticated identity — not
 * automation), and deferring PRs locally.
 */

import type { AragoraClient } from '../client';

/** Brief verdict enum. */
export type BriefVerdict =
  | 'approve_candidate'
  | 'needs_human_attention'
  | 'repair_first';

/** PR entry in the review queue list response. */
export interface ReviewQueuePR {
  number: number;
  title: string;
  author: string;
  head_sha: string;
  created_at: string;
  subsystems: string[];
  check_counts: Record<string, number>;
  has_brief: boolean;
  verdict?: BriefVerdict;
}

/** Session statistics for the header badge. */
export interface ReviewQueueStats {
  approved_today: number;
  median_decision_seconds?: number;
  streak: number;
}

/**
 * Review Queue namespace for browser-based PR settlement triage.
 *
 * @example
 * ```typescript
 * const { prs } = await client.reviewQueue.listPRs();
 * for (const pr of prs) {
 *   console.log(pr.number, pr.title);
 * }
 * ```
 */
export class ReviewQueueAPI {
  constructor(private client: AragoraClient) {}

  /** List open PRs currently in the review queue. */
  async listPRs(options?: {
    includeDeferred?: boolean;
  }): Promise<{ prs: ReviewQueuePR[]; total: number }> {
    const params: Record<string, unknown> = {};
    if (options?.includeDeferred) {
      params.include_deferred = '1';
    }
    return this.client.request<{ prs: ReviewQueuePR[]; total: number }>(
      'GET',
      '/api/v1/review-queue/prs',
      { params }
    );
  }

  /** Fetch the PDB brief for a specific PR (404 if no brief exists). */
  async getBrief(prNumber: number): Promise<{ brief: Record<string, unknown> }> {
    return this.client.request<{ brief: Record<string, unknown> }>(
      'GET',
      `/api/v1/review-queue/prs/${prNumber}/brief`
    );
  }

  /** Submit a GitHub APPROVE review for the PR using the caller's identity. */
  async approve(
    prNumber: number,
    options?: { note?: string }
  ): Promise<{ ok: boolean; review_id?: number }> {
    const body: Record<string, unknown> = {};
    if (options?.note !== undefined) {
      body.note = options.note;
    }
    return this.client.request<{ ok: boolean; review_id?: number }>(
      'POST',
      `/api/v1/review-queue/prs/${prNumber}/approve`,
      { body }
    );
  }

  /** Submit a REQUEST_CHANGES review with a required reason. */
  async requestChanges(
    prNumber: number,
    reason: string
  ): Promise<{ ok: boolean; review_id?: number }> {
    return this.client.request<{ ok: boolean; review_id?: number }>(
      'POST',
      `/api/v1/review-queue/prs/${prNumber}/request-changes`,
      { body: { reason } }
    );
  }

  /** Defer the PR locally (hides it from the queue for ~4 hours). */
  async defer(prNumber: number): Promise<{ ok: boolean; deferred_until: string }> {
    return this.client.request<{ ok: boolean; deferred_until: string }>(
      'POST',
      `/api/v1/review-queue/prs/${prNumber}/defer`
    );
  }

  /** Fetch session stats for the header badge. */
  async stats(): Promise<ReviewQueueStats> {
    return this.client.request<ReviewQueueStats>(
      'GET',
      '/api/v1/review-queue/stats'
    );
  }
}
